File: lib/graph/port.c
/*
 * Copyright 2017-2018 Philippe Proulx <pproulx@efficios.com>
 * Copyright 2017 Jérémie Galarneau <jeremie.galarneau@efficios.com>
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

#define BT_LOG_TAG "PORT"
#include <babeltrace/lib-logging-internal.h>

#include <babeltrace/assert-internal.h>
#include <babeltrace/assert-pre-internal.h>
#include <babeltrace/graph/port-const.h>
#include <babeltrace/graph/port-input-const.h>
#include <babeltrace/graph/port-output-const.h>
#include <babeltrace/graph/self-component-port.h>
#include <babeltrace/graph/self-component-port-input.h>
#include <babeltrace/graph/self-component-port-output.h>
#include <babeltrace/graph/component-internal.h>
#include <babeltrace/graph/port-internal.h>
#include <babeltrace/graph/connection-internal.h>
#include <babeltrace/object-internal.h>
#include <babeltrace/compiler-internal.h>

static
void destroy_port(struct bt_object *obj)
{
	struct bt_port *port = (void *) obj;

	BT_LIB_LOGD("Destroying port: %!+p", port);

	if (port->name) {
		g_string_free(port->name, TRUE);
		port->name = NULL;
	}

	g_free(port);
}

BT_HIDDEN
struct bt_port *bt_port_create(struct bt_component *parent_component,
		enum bt_port_type type, const char *name, void *user_data)
{
	struct bt_port *port = NULL;

	BT_ASSERT(name);
	BT_ASSERT(parent_component);
	BT_ASSERT(type == BT_PORT_TYPE_INPUT || type == BT_PORT_TYPE_OUTPUT);
	BT_ASSERT(strlen(name) > 0);
	port = g_new0(struct bt_port, 1);
	if (!port) {
		BT_LOGE_STR("Failed to allocate one port.");
		goto end;
	}

	BT_LIB_LOGD("Creating port for component: %![comp-]+c, port-type=%s, "
		"port-name=\"%s\"", parent_component, bt_port_type_string(type),
		name);
	bt_object_init_shared_with_parent(&port->base, destroy_port);
	port->name = g_string_new(name);
	if (!port->name) {
		BT_LOGE_STR("Failed to allocate one GString.");
		BT_OBJECT_PUT_REF_AND_RESET(port);
		goto end;
	}

	port->type = type;
	port->user_data = user_data;
	bt_object_set_parent(&port->base, &parent_component->base);
	BT_LIB_LOGD("Created port for component: "
		"%![comp-]+c, %![port-]+p", parent_component, port);

end:
	return port;
}

const char *bt_port_get_name(const struct bt_port *port)
{
	BT_ASSERT_PRE_NON_NULL(port, "Port");
	return port->name->str;
}

enum bt_port_type bt_port_get_type(const struct bt_port *port)
{
	BT_ASSERT_PRE_NON_NULL(port, "Port");
	return port->type;
}

const struct bt_connection *bt_port_borrow_connection_const(
		const struct bt_port *port)
{
	BT_ASSERT_PRE_NON_NULL(port, "Port");
	return port->connection;
}

const struct bt_component *bt_port_borrow_component_const(
		const struct bt_port *port)
{
	BT_ASSERT_PRE_NON_NULL(port, "Port");
	return bt_port_borrow_component_inline(port);
}

struct bt_self_component *bt_self_component_port_borrow_component(
		struct bt_self_component_port *port)
{
	BT_ASSERT_PRE_NON_NULL(port, "Port");
	return (void *) bt_object_borrow_parent((void *) port);
}

BT_HIDDEN
void bt_port_set_connection(struct bt_port *port,
		struct bt_connection *connection)
{
	/*
	 * Don't take a reference on connection as its existence is
	 * guaranteed by the existence of the graph in which the
	 * connection exists.
	 */
	port->connection = connection;
	BT_LIB_LOGV("Set port's connection: %![port-]+p, %![conn-]+x", port,
		connection);
}

static inline
bool port_connection_iterators_are_finalized(struct bt_port *port)
{
	bool ret = true;
	struct bt_connection *conn = port->connection;
	uint64_t i;

	if (!conn) {
		goto end;
	}

	for (i = 0; i < conn->iterators->len; i++) {
		struct bt_self_component_port_input_message_iterator *iterator =
			conn->iterators->pdata[i];

		BT_ASSERT(iterator);

		if (iterator->state != BT_SELF_COMPONENT_PORT_INPUT_MESSAGE_ITERATOR_STATE_FINALIZING &&
				iterator->state != BT_SELF_COMPONENT_PORT_INPUT_MESSAGE_ITERATOR_STATE_FINALIZED) {
			BT_ASSERT_PRE_MSG("Message iterator is not being finalized or finalized: "
				"%!+i", iterator);
			ret = false;
			goto end;
		}
	}

end:
	return ret;
}

bt_bool bt_port_is_connected(const struct bt_port *port)
{
	BT_ASSERT_PRE_NON_NULL(port, "Port");
	return port->connection ? BT_TRUE : BT_FALSE;
}

void *bt_self_component_port_get_data(const struct bt_self_component_port *port)
{
	BT_ASSERT_PRE_NON_NULL(port, "Port");
	return ((struct bt_port *) port)->user_data;
}

void bt_port_get_ref(const struct bt_port *port)
{
	bt_object_get_ref(port);
}

void bt_port_put_ref(const struct bt_port *port)
{
	bt_object_put_ref(port);
}

void bt_port_input_get_ref(const struct bt_port_input *port_input)
{
	bt_object_get_ref(port_input);
}

void bt_port_input_put_ref(const struct bt_port_input *port_input)
{
	bt_object_put_ref(port_input);
}

void bt_port_output_get_ref(const struct bt_port_output *port_output)
{
	bt_object_get_ref(port_output);
}

void bt_port_output_put_ref(const struct bt_port_output *port_output)
{
	bt_object_put_ref(port_output);
}

File: bindings/python/bt2/bt2/port.py
# The MIT License (MIT)
#
# Copyright (c) 2017 Philippe Proulx <pproulx@efficios.com>
#
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in
# all copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
# THE SOFTWARE.

from bt2 import native_bt, object, utils
import collections.abc
import bt2.component
import bt2.connection
import bt2.message_iterator
import bt2.message
import copy
import bt2


def _create_from_ptr(ptr):
    port_type = native_bt.port_get_type(ptr)

    if port_type == native_bt.PORT_TYPE_INPUT:
        cls = _InputPort
    elif port_type == native_bt.PORT_TYPE_OUTPUT:
        cls = _OutputPort
    else:
        raise bt2.Error('unknown port type: {}'.format(port_type))

    return cls._create_from_ptr(ptr)


def _create_private_from_ptr(ptr):
    pub_ptr = native_bt.port_from_private(ptr)
    utils._handle_ptr(pub_ptr, 'cannot get port object from private port object')
    port_type = native_bt.port_get_type(pub_ptr)
    assert(port_type == native_bt.PORT_TYPE_INPUT or port_type == native_bt.PORT_TYPE_OUTPUT)

    if port_type == native_bt.PORT_TYPE_INPUT:
        cls = _PrivateInputPort
    elif port_type == native_bt.PORT_TYPE_OUTPUT:
        cls = _PrivateOutputPort

    obj = cls._create_from_ptr(ptr)
    obj._pub_ptr = pub_ptr
    return obj


class _Port(object._SharedObject):
    @staticmethod
    def _name(ptr):
        name = native_bt.port_get_name(ptr)
        assert(name is not None)
        return name

    @staticmethod
    def _disconnect(ptr):
        status = native_bt.port_disconnect(ptr)

        if status < 0:
            raise bt2.Error('cannot disconnect port')

    @property
    def name(self):
        return self._name(self._ptr)

    @property
    def component(self):
        comp_ptr = native_bt.port_get_component(self._ptr)

        if comp_ptr is None:
            return

        return bt2.component._create_generic_component_from_ptr(comp_ptr)

    @property
    def connection(self):
        conn_ptr = native_bt.port_get_connection(self._ptr)

        if conn_ptr is None:
            return

        return bt2.connection._Connection._create_from_ptr(conn_ptr)

    @property
    def is_connected(self):
        return self.connection is not None

    def disconnect(self):
        self._disconnect(self._ptr)

    def __eq__(self, other):
        if type(other) is not type(self):
            return False

        return self.addr == other.addr


class _InputPort(_Port):
    pass


class _OutputPort(_Port):
    def create_message_iterator(self, message_types=None,
                                     colander_component_name=None):
        msg_types = bt2.message._msg_types_from_msg_classes(message_types)

        if colander_component_name is not None:
            utils._check_str(colander_component_name)

        msg_iter_ptr = native_bt.py3_create_output_port_msg_iter(int(self._ptr),
                                                                     colander_component_name,
                                                                     msg_types)

        if msg_iter_ptr is None:
            raise bt2.CreationError('cannot create output port message iterator')

        return bt2.message_iterator._OutputPortMessageIterator._create_from_ptr(msg_iter_ptr)


class _PrivatePort(_Port):
    @property
    def name(self):
        return self._name(self._pub_ptr)

    @property
    def component(self):
        comp_ptr = native_bt.private_port_get_private_component(self._ptr)

        if comp_ptr is None:
            return

        pub_comp_ptr = native_bt.component_from_private(comp_ptr)
        assert(pub_comp_ptr)
        comp = bt2.component._create_generic_component_from_ptr(pub_comp_ptr)
        native_bt.put(comp_ptr)
        return comp

    @property
    def connection(self):
        conn_ptr = native_bt.private_port_get_private_connection(self._ptr)

        if conn_ptr is None:
            return

        return bt2.connection._create_private_from_ptr(conn_ptr)

    def remove_from_component(self):
        status = native_bt.private_port_remove_from_component(self._ptr)

        if status < 0:
            raise bt2.Error("cannot remove port from component")

    def disconnect(self):
        self._disconnect(self._pub_ptr)


class _PrivateInputPort(_PrivatePort, _InputPort):
    pass


class _PrivateOutputPort(_PrivatePort, _OutputPort):
    pass
